### ISSUE TITLE
Phase 63.2: DepthEstimator — Monocular, Stereo & Multi-View Depth Estimation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1179]
### Overview
The DepthEstimator provides monocular, stereo, and multi-view depth estimation capabilities using state-of-the-art neural networks. It implements MiDaS, DPT, and Depth Anything architectures for single-image depth prediction, along with stereo matching and multi-view stereo (MVSNet) for geometry-based depth recovery. The module handles metric depth recovery, depth completion from sparse inputs, and depth-aware feature extraction for downstream 3D tasks.

### Architecture

```mermaid
graph TB
    subgraph DepthEstimator
        IN[ImageInput] --> MON[MonocularEstimator]
        IN --> ST[StereoMatcher]
        IN --> MVS[MultiViewStereo]
        
        subgraph MonocularDepth
            MON --> MIDAS[MiDaS]
            MON --> DPT_M[DPT]
            MON --> DA[DepthAnything]
        end
        
        subgraph StereoDepth
            ST --> SGBM[SemiGlobalMatching]
            ST --> RAFT_S[RAFTStereo]
            ST --> CRE[CREStereo]
        end
        
        subgraph MVSDepth
            MVS --> MVSN[MVSNet]
            MVS --> CAS[CascadeMVS]
        end
        
        MON --> DC[DepthCompletion]
        ST --> DC
        MVS --> DC
        DC --> MET[MetricRecovery]
        MET --> FEAT[DepthFeatureExtractor]
    end
```

### Interface Definitions

```typescript
interface DepthMap {
  data: Float32Array;         // [H * W] depth values
  width: number;
  height: number;
  minDepth: number;
  maxDepth: number;
  isMetric: boolean;          // true if values are in meters
  confidence?: Float32Array;  // [H * W] per-pixel confidence
  format: 'inverse_depth' | 'linear_depth' | 'log_depth';
}

interface MonocularDepthConfig {
  model: 'midas_v3' | 'midas_v3.1' | 'dpt_large' | 'dpt_hybrid' | 'depth_anything_v2' | 'depth_anything_v2_metric';
  inputResolution: [number, number];
  outputResolution?: [number, number];
  useMetricDepth: boolean;
  ensembleSteps?: number;      // for multi-pass refinement
  backbone: 'vitl' | 'vitb' | 'vits' | 'efficientnet';
}

interface StereoConfig {
  method: 'sgbm' | 'raft_stereo' | 'crestereo' | 'igev';
  maxDisparity: number;
  baseline: number;             // camera baseline in meters
  focalLength: number;          // focal length in pixels
  confidenceThreshold: number;
}

interface MultiViewStereoConfig {
  method: 'mvsnet' | 'cascade_mvsnet' | 'patchmatchnet';
  numDepthPlanes: number;
  depthRange: [number, number];
  numSourceViews: number;
  fusionMethod: 'tsdf' | 'poisson' | 'point_cloud';
}

interface DepthCompletionConfig {
  sparseInput: Float32Array;   // known depth values (0 = unknown)
  guidanceImage?: ImageData;   // RGB guidance for completion
  method: 'ip_basic' | 'nconv' | 'guided_diffusion';
}

interface CameraIntrinsics {
  fx: number; fy: number;
  cx: number; cy: number;
  width: number; height: number;
  distortion?: number[];
}

interface DepthEstimator {
  // Monocular estimation
  estimateMonocular(image: ImageData, config: MonocularDepthConfig): Promise<DepthMap>;
  estimateRelativeDepth(image: ImageData): Promise<DepthMap>;
  estimateMetricDepth(image: ImageData, intrinsics?: CameraIntrinsics): Promise<DepthMap>;
  
  // Stereo estimation
  estimateStereo(left: ImageData, right: ImageData, config: StereoConfig): Promise<DepthMap>;
  rectifyStereoPair(left: ImageData, right: ImageData, calibration: StereoCalibration): Promise<[ImageData, ImageData]>;
  
  // Multi-view stereo
  estimateMultiView(images: ImageData[], cameras: CameraPose[], config: MultiViewStereoConfig): Promise<DepthMap[]>;
  fuseMVSDepthMaps(depthMaps: DepthMap[], cameras: CameraPose[]): Promise<PointCloud>;
  
  // Depth completion
  completeSparseDepth(config: DepthCompletionConfig): Promise<DepthMap>;
  
  // Utilities
  depthToPointCloud(depth: DepthMap, intrinsics: CameraIntrinsics, color?: ImageData): PointCloud;
  depthToNormalMap(depth: DepthMap, intrinsics: CameraIntrinsics): Float32Array;
  alignScaleAndShift(predicted: DepthMap, groundTruth: DepthMap): DepthMap;
  evaluateDepth(predicted: DepthMap, groundTruth: DepthMap): DepthMetrics;
}

interface DepthMetrics {
  absRel: number;    // Absolute Relative Error
  sqRel: number;     // Squared Relative Error
  rmse: number;      // Root Mean Square Error
  rmseLog: number;   // RMSE of log depth
  delta1: number;    // % of pixels with max(d/d*, d*/d) < 1.25
  delta2: number;    // threshold 1.25^2
  delta3: number;    // threshold 1.25^3
  siLog: number;     // Scale-Invariant Log Error
}
```

### Key Components

1. **MiDaSEstimator** — Multi-dataset trained monocular depth with zero-shot cross-dataset transfer, supporting v3 and v3.1 architectures
2. **DPTEstimator** — Vision Transformer backbone (ViT-Large/Hybrid) for dense prediction with reassemble and fusion stages
3. **DepthAnythingEstimator** — Large-scale self-supervised pretraining with unlabeled data, supporting both relative and metric depth variants
4. **StereoMatcher** — Semi-global matching (SGBM) and learned stereo (RAFT-Stereo, CREStereo) with confidence estimation
5. **MVSNetPipeline** — Cost volume-based multi-view stereo with plane-sweep hypothesis testing and depth map fusion
6. **DepthCompletor** — Sparse-to-dense depth completion with RGB guidance using normalized convolutions
7. **MetricDepthRecovery** — Scale and shift alignment for converting relative depth to metric depth using sparse anchor points
8. **DepthPointCloudProjector** — Pinhole camera model depth-to-3D projection with distortion handling

### Testing Strategy
- Unit tests for each monocular depth model with reference images and expected depth ranges
- Stereo matching tests with Middlebury stereo benchmark pairs
- Multi-view stereo tests with DTU/Tanks-and-Temples subsets
- Depth completion tests measuring accuracy vs. density of sparse input
- Metric depth recovery tests with known ground-truth LiDAR scans
- Point cloud projection tests verifying geometric consistency
- Performance benchmarks: 640x480 monocular depth in <50ms on GPU

### References
- Ranftl et al. — *Towards Robust Monocular Depth Estimation: Mixing Datasets for Zero-Shot Cross-Dataset Transfer* (T-PAMI 2020) [MiDaS]
- Ranftl et al. — *Vision Transformers for Dense Prediction* (ICCV 2021) [DPT]
- Yang et al. — *Depth Anything: Unleashing the Power of Large-Scale Unlabeled Data* (CVPR 2024)
- Yao et al. — *MVSNet: Depth Inference for Unstructured Multi-view Stereo* (ECCV 2018)
- Teed & Deng — *RAFT-Stereo: Multilevel Recurrent Field Transforms for Stereo Matching* (3DV 2021)
- Li et al. — *Practical Stereo Matching via Cascaded Recurrent Network with Adaptive Correlation* (CVPR 2022) [CREStereo]
- Hirschmuller — *Stereo Processing by Semi-Global Matching and Mutual Information* (T-PAMI 2008) [SGM]
